webrtc/RTCPeerConnection-setRemoteDescription.html

```diff
@@ -48,76 +45,101 @@
   /*
     4.6.1.  enum RTCSdpType
    */
-  promise_test(t => {
+  promise_test(async t => {
     const pc = new RTCPeerConnection();
+    t.add_cleanup(() => pc.close());
 
     // SDP is validated after WebIDL validation
-    return promise_rejects(t, new TypeError(),
-      pc.setRemoteDescription({
-        type: 'bogus',
-        sdp: 'bogus'
-      }));
+    try {
+      await pc.setRemoteDescription({ type: 'bogus', sdp: 'bogus' });
+      t.unreached_func("Should have rejected.");
+    } catch (e) {
+      assert_throws(new TypeError(), () => { throw e });
+    }
   }, 'setRemoteDescription with invalid type and invalid SDP should reject with TypeError');
 
-  promise_test(t => {
+  promise_test(async t => {
     const pc = new RTCPeerConnection();
+    t.add_cleanup(() => pc.close());
 
     // SDP is validated after validating type
-    return promise_rejects(t, 'InvalidStateError',
-      pc.setRemoteDescription({
-        type: 'answer',
-        sdp: 'invalid'
-      }));
+    try {
+      await pc.setRemoteDescription({ type: 'answer', sdp: 'invalid' });
+      t.unreached_func("Should have rejected.");
+    } catch (e) {
+      assert_throws('InvalidStateError', () => { throw e });
+    }
   }, 'setRemoteDescription() with invalid SDP and stable state should reject with InvalidStateError');
 
-  /* Operations after returning to stable state */
+  /* Dedicated signalingstate events test. */
 
-  promise_test(t => {
+  promise_test(async t => {
     const pc = new RTCPeerConnection();
     const pc2 = new RTCPeerConnection();
+    t.add_cleanup(() => pc.close());
+    t.add_cleanup(() => pc2.close());
+
+    let eventCount = 0;
+    const states = [
+      'stable', 'have-local-offer', 'stable', 'have-remote-offer', 'closed'
+    ];
+    pc.onsignalingstatechange = () =>
```

Comment:
You need to wrap this callback in t.step_func so that any exception is linked with the test rather than hitting the global error handler. i.e. this should read

```
    pc.onsignalingstatechange = t.step_func(() =>
        assert_equals(pc.signalingState, states[++eventCount]));
```

With that change `wpt run --verify chrome webrtc/RTCPeerConnection-setRemoteDescription.html` passes locally with the branch checked out. You can either push the fixup to the PR branch or to gecko, whichever is easiest.

Reply:
@jgraham should that be done for any event handler?

I just ran into an issue with [exchangeIceCandidates](https://github.com/web-platform-tests/wpt/blob/master/webrtc/RTCPeerConnection-helper.js#L222) throwing errors, probably when called *after* the test ended and the cleanup was executed.
If using t.step_func will also not execute after the test ends that would be even better :-)

Reply:
Yes, anything that's a callback needs to be wrapped in `step_func` unless you only have one test on the page in which case you can use a single page test. However that won't have the effect you want in terms of the function not executing after the test is done, whereas a step function will terminate early in that case (iirc).